overlord/restart/restart.go

```diff
@@ -137,6 +193,47 @@ func (rm *RestartManager) rebootDidNotHappen() error {
 	return nil
 }
 
+// PendingForSystemRestart returns true if the change has tasks that are held
+// pending a manual system restart. It is registered with the prune logic.
+func (rm *RestartManager) PendingForSystemRestart(chg *state.Change) bool {
```

Comment:
It seems that most of this code is the same as in StartUp(). You could move the common parts into a helper function
```go
enumerateTasksPendingOnRestart(chg *state.Change, func(task *state.Task) bool)
```
that would invoke a callback for every task who was pending on this restart, and could be shortcut if the callback returning `false` (meaning it's not interested in enumerating further), so that you could implement `PendingForSystemRestart` by something like this:
```go
func (rm *RestartManager) PendingForSystemRestart(chg *state.Change) bool {
    foundPendingTask := false
    enumerateTasksPendingOnRestart(chg, func(task *state.Task) bool {
        foundPendingTask = true
        return false
    })
    return foundPendingTask
}
```

Reply:
I tried this in the world after the fix in https://github.com/snapcore/snapd/pull/12334, it doesn't save lines and the code is a bit less readable, the issue is also that after the fix there is less shared code because of the different condition on the boot ids.